metrics/config.go

```diff
@@ -100,10 +102,62 @@ type metricsConfig struct {
 	// E.g., "custom.googleapis.com/<subdomain>/<component>".
 	// Store this in a variable to reduce string join operations.
 	stackdriverCustomMetricTypePrefix string
+	// stackdriverClientConfig is the metadata to configure the metrics exporter's Stackdriver client.
+	stackdriverClientConfig stackdriverClientConfig
+}
+
+// stackdriverClientConfig encapsulates the metadata required to configure a Stackdriver client.
+type stackdriverClientConfig struct {
+	// ProjectID is the stackdriver project ID to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ProjectID string
+	// GCPLocation is the GCP region or zone to which data is uploaded.
+	// This is not necessarily the GCP location where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	GCPLocation string
+	// ClusterName is the cluster name with which the data will be associated in Stackdriver.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ClusterName string
+	// GCPSecretName is the optional GCP service account key which will be used to
+	// authenticate with Stackdriver. If not provided, Google Application Default Credentials
+	// will be used (https://cloud.google.com/docs/authentication/production).
+	GCPSecretName string
+	// GCPSecretNamespace is the Kubernetes namespace where GCPSecretName is located.
+	// The Kubernetes ServiceAccount used by the pod that is exporting data to
+	// Stackdriver should have access to Secrets in this namespace.
+	GCPSecretNamespace string
+}
+
+// newStackdriverClientConfigFromMap creates a stackdriverClientConfig from the given map
+func newStackdriverClientConfigFromMap(config map[string]string) (*stackdriverClientConfig, error) {
+	sc := &stackdriverClientConfig{}
+
+	if pi, ok := config[StackdriverProjectIDKey]; ok {
+		sc.ProjectID = pi
+	}
+
+	if gl, ok := config[StackdriverGCPLocationKey]; ok {
+		sc.GCPLocation = gl
+	}
+
+	if cn, ok := config[StackdriverClusterNameKey]; ok {
+		sc.ClusterName = cn
+	}
+
+	if gsn, ok := config[StackdriverGCPSecretNameKey]; ok {
+		sc.GCPSecretName = gsn
+	}
+
+	if gsns, ok := config[StackdriverGCPSecretNamespaceKey]; ok {
+		sc.GCPSecretNamespace = gsns
+	}
```

Comment:
Since maps return nil value if the key doesn't exist, in this case (strings and no default values),
```
...
sc.GCPSecretNamespace = config[StackdriverGCPSecretNamespaceKey]
...
```
would be enough.

Reply:
Nice, will change.